### DISCUSSION TITLE
bug in build script on centos 7?

### DISCUSSION THREAD

**Discussion by jgaer** [atom/atom#22568] [Q&A]
-jgaer@ljgaer2_~/atom-master: script/build --create-rpm-package
/home/jgaer/atom-master/script/build:95
 .then(async packagedAppPath => {
 ^^^^^

SyntaxError: missing ) after argument list  at createScript (vm.js:56:10)  at Object.runInThisContext (vm.js:97:10)  at Module._compile (module.js:549:28)  at Object.Module._extensions..js (module.js:586:10)  at Module.load (module.js:494:32)  at tryModuleLoad (module.js:453:12)  at Function.Module._load (module.js:445:3)  at Module.runMain (module.js:611:10)  at run (bootstrap_node.js:394:7)  at startup (bootstrap_node.js:160:9)
jgaer@ljgaer2_~/atom-master:

**Comment by ThatXliner** [comment]
What's your version of Node? (`node --version`)

**Reply by jgaer** [reply]
6.17.1

**Reply by sadick254** [reply]
Try upgrading to node version 12 and then try to build again.

**Reply by jgaer** [reply]
I was able to get nodejs 12.16.3  installed from a tarball. I still can't build atom. It looks like I need access to a git repository, but I am on a disconnected network. Or do I just need to set GIT_DISCOVERY_ACROSS_FILESYSTEM to some value like TRUE?

%script/build

fatal: not a git repository (or any parent up to mount point /home) Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

/home/jgaer/atom-master/script/lib/spawn-sync.js:17
 `Command ${result.args.join(' ')} exited with code "${result.status}"`  ^

TypeError: Cannot read property 'join' of undefined  at module.exports (/home/jgaer/atom-master/script/lib/spawn-sync.js:17:30)
 at computeAppVersion (/home/jgaer/atom-master/script/config.js:95:20)
 at Object.<anonymous> (/home/jgaer/atom-master/script/config.js:24:28)
 at Module._compile (internal/modules/cjs/loader.js:1133:30)
 at Object.Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
 at Module.load (internal/modules/cjs/loader.js:977:32)
 at Function.Module._load (internal/modules/cjs/loader.js:877:14)
 at Module.require (internal/modules/cjs/loader.js:1019:19)
 at require (internal/modules/cjs/helpers.js:77:18)
 at Object.<anonymous> (/home/jgaer/atom-master/script/bootstrap:6:16)

**Reply by DeeDeeG** [reply]
It looks like you may have downloaded the Atom repository as a zip archive or something like that.

If it's possible for you to get it as a git repository, that would be great. (If you really need it as a .zip, you can clone github.com/atom/atom, perhaps as a bare clone if you want/if you know how to do that, then however you cloned it you can zip that up and get it onto the machine you intend to build on. Then extract.)

Parts of the build script appear use git tags to determine what version of Atom is being built, but this will fail if you aren't in a directory that is a real git repository for the Atom project, with the commits and tags fetched and all.

If all that is too hard, but you are handy with JavaScript, you can edit `script/config`; Figure out what the result of `computeAppVersion()` would be (or make a guess) and update the function to simply return a hard-coded string rather tan running `git` in a subprocess...

I'm not sure if Atom will build on CentOS 7, but it's worth a try.

**Reply by jgaer** [reply]
well hardcoding the version got me past that error. But we are on an isolated network ( no internet connectivity) and now it looks like the build is hung trying to pull some dependencies in. The last message I get is "Installing script dependencies".  Does the build script produce an RPM. I am thinking that if I can set up a VM and install Redhat7.7 on it outside on a network connected machine I could move the RPM over into the isolated network and install. The rpm currently available requires a different version of glibc than is installed on my box.

**Reply by DeeDeeG** [reply]
Yes, the build script does produce an rpm package when run with the `--create-rpm-package` flag as you did. It pulls in lots of dependencies.

I think one of said dependencies uses some newer C or C++ features, relying on a build toolchain new enough to supports those newer language features. I forget which, off the top of my head. If you can get that to build somehow, without it becoming too new to finally run on an unmodified CentOS 7 image... Then in theory it could work. I think you might end up with something that still requires too new of a standard C library `glibc`, though. If it does work, let us know.

**Reply by jgaer** [reply]
Definitely requires network access to build. Build script eventually returned -

script/build
Node: v12.16.3
Npm: v6.14.4
Python: v2.7.5
Installing script dependencies
npm ERR! Linux 3.10.0-1062.4.3.el7.x86_64 npm ERR! argv "/usr/bin/node" "/bin/npm" "--loglevel=error" "install"
npm ERR! node v6.17.1
npm ERR! npm v3.10.10
npm ERR! code ETIMEDOUT
npm ERR! errno ETIMEDOUT
npm ERR! syscall connect

npm ERR! network connect ETIMEDOUT 104.16.16.35:443 npm ERR! network This is most likely not a problem with npm itself npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the npm ERR! network 'proxy' config is set properly. See: 'npm help config'

npm ERR! Please include the following file with any support request:
npm ERR! /home/jgaer/atom-master/script/npm-debug.log
child_process.js:651
 throw err;
 ^

Error: Command failed: npm --loglevel=error install npm ERR! Linux 3.10.0-1062.4.3.el7.x86_64 npm ERR! argv "/usr/bin/node" "/bin/npm" "--loglevel=error" "install"
npm ERR! node v6.17.1
npm ERR! npm v3.10.10
npm ERR! code ETIMEDOUT
npm ERR! errno ETIMEDOUT
npm ERR! syscall connect

npm ERR! network connect ETIMEDOUT 104.16.16.35:443 npm ERR! network This is most likely not a problem with npm itself npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the npm ERR! network 'proxy' config is set properly. See: 'npm help config'

npm ERR! Please include the following file with any support request:
npm ERR! /home/jgaer/atom-master/script/npm-debug.log

at checkExecSyncError (child_process.js:630:11)  at Object.execFileSync (child_process.js:648:15)  at module.exports (/home/jgaer/atom-master/script/lib/install-script-dependencies.js:12:16)
 at Object.<anonymous> (/home/jgaer/atom-master/script/bootstrap:37:1)
 at Module._compile (internal/modules/cjs/loader.js:1133:30)
 at Object.Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
 at Module.load (internal/modules/cjs/loader.js:977:32)
 at Function.Module._load (internal/modules/cjs/loader.js:877:14)
 at Module.require (internal/modules/cjs/loader.js:1019:19)
 at require (internal/modules/cjs/helpers.js:77:18) {
 status: 1,
 signal: null,
 output: [
 null,
 <Buffer >,
 <Buffer 6e 70 6d 20 45 52 52 21 20 4c 69 6e 75 78 20 33 2e 31 30 2e 30 2d 31 30 36 32 2e 34 2e 33 2e 65 6c 37 2e 78 38 36 5f 36 34 0a 6e 70 6d 20 45 52 52 21 ... 731 more bytes>  ],
 pid: 3888,
 stdout: <Buffer >,
 stderr: <Buffer 6e 70 6d 20 45 52 52 21 20 4c 69 6e 75 78 20 33 2e 31 30 2e 30 2d 31 30 36 32 2e 34 2e 33 2e 65 6c 37 2e 78 38 36 5f 36 34 0a 6e 70 6d 20 45 52 52 21 ... 731 more bytes> }

**Reply by DeeDeeG** [reply]
I am seeing that the superstring (https://github.com/atom/superstring) dependency doesn't build on CentOS 7.

Looks like it's using newer c++ language features that are  not supported with the old compiler toolchain on CentOS 7?

```
../src/core/text-buffer.cc: In constructor 'TextBuffer::Snapshot::Snapshot(TextBuffer&, TextBuffer::Layer&, TextBuffer::Layer&)':
../src/core/text-buffer.cc:1073:56: error: invalid initialization of non-const reference of type 'TextBuffer&' from an rvalue of type '<brace-enclosed initializer list>'
   : buffer{buffer}, layer{layer}, base_layer{base_layer} {}
                                                        ^
../src/core/text-buffer.cc:1073:56: error: invalid initialization of non-const reference of type 'TextBuffer::Layer&' from an rvalue of type '<brace-enclosed initializer list>'
../src/core/text-buffer.cc:1073:56: error: invalid initialization of non-const reference of type 'TextBuffer::Layer&' from an rvalue of type '<brace-enclosed initializer list>'
make: *** [Release/obj.target/superstring_core/src/core/text-buffer.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/root/atom/apm/node_modules/atom-package-manager/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:223:5)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
gyp ERR! System Linux 5.4.0-74-generic
gyp ERR! command "/root/atom/apm/node_modules/atom-package-manager/bin/node" "/root/atom/apm/node_modules/atom-package-manager/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /root/atom/node_modules/superstring
gyp ERR! node -v v12.14.1
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! superstring@2.4.4 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the superstring@2.4.4 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.atom/.apm/_logs/2021-06-23T13_07_04_066Z-debug.log
```

**Reply by DeeDeeG** [reply]
Although if you can get newer GCC/G++ (version 4.8.1 or newer) installed and running, the compilation error may go away?

- https://flight-manual.atom.io/hacking-atom/sections/hacking-on-atom-core/
  - > If `script/build` exits with an error, you may need to install a newer C++ compiler with C++11:
- https://gcc.gnu.org/projects/cxx-status.html#cxx11
  - > GCC 4.8.1 was the first feature-complete implementation of the 2011 C++ standard, previously known as C++0x.

**Reply by DeeDeeG** [reply]
I see GCC 4.8.5 is what is installed in my CentOS Docker container ... and it's still giving me that error? And it also fails with that error on Ubuntu 14.04 "Trusty" with Clang 5... :shrug: I'm not sure what it takes. I just know it builds on the current Ubuntu and Fedora distros, at least.

**Reply by VanshajPoonia** [reply]
You must have *git* installed on your system.
Now, go to the folder where you want to have your Atom editor and then press right-click and then select the option `Git bash Here`. Now, fork the atom repo and then use the command `git clone git@github.com:your-username/atom.git` and replace _your-username_ with your GitHub Username. Now, just do `cd atom` and after that, run `script/bootstrap`. It worked in my case at least.

**Comment by VanshajPoonia** [comment]
Use git bash for running the commands. Worked for me.
